### ISSUE TITLE
Use direct I/O for loop devices

### ISSUE THREAD

**Issue by DemiMarie** [QubesOS/qubes-issues#7332]
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### The problem you're addressing (if any)

Qubes OS currently does not use direct I/O for the pools that use loop devices  This leads to double-caching and poor I/O performance.  I believe it is the cause of the anomalies [in this benchmark](https://github.com/QubesOS/qubes-issues/issues/7300#issuecomment-1063035390).

### The solution you'd like

Use O_DIRECT for loop devices.

### The value to a user, and who that user might be

Users of the reflink pool will experience improved and more consistent performance.

**Comment by AlxHnr** [comment]
Could this be causing #5086?

**Comment by DemiMarie** [comment]
Possibly

**Comment by DemiMarie** [comment]
Reopening as QubesOS/qubes-vmm-xen#127 had to be reverted (#7828).

**Comment by marmarek** [comment]
For the record, the next iteration of this patch (if it comes), I'm going to block on actually proving it helps with performance. We see it's more risky change than it seems to be, so lets be sure it's actually worth doing.

**Comment by DemiMarie** [comment]
I agree.  Someone on the forum seems to have a benchmark setup that could be used.

**Comment by rustybird** [comment]
If someone wants to play around with this on `file-reflink`  without having to build a bleeding edge util-linux `losetup`, here's a small Python [`piccolosetup`](https://gist.github.com/rustybird/ff6ba6a9b583b93624746df862dd2cb7) that's good enough to do one proper `LOOP_CONFIGURE` ioctl from a [patched](https://gist.github.com/rustybird/21f0cdde2f84c4f3146333896f25c6ac) `/etc/xen/scripts/block`.

    Usage: piccolosetup [--sector-size=BYTES] [--direct-io=on] [-r] LOOPDEV FILE

**Comment by brendanhoar** [comment]
Looks like #7828 now includes a clarification that the failure of the first iteration of the patch (on 4Kn drives) applied to LVM thick, not LVM thin. For the record.

B

**Comment by rustybird** [comment]
In the LVM Thin installation layout it would still have affected the legacy 'file' driver `varlibqubes` pool on the ext4 dom0 filesystem. (And the lvm_thin driver `vm` pool would be broken anyway on a 4Kn drive.)
